processor/src/trace.rs

```diff
@@ -142,7 +195,123 @@ impl<'a> TraceFragment<'a> {
+            2 => {
+                // initialize hasher and bitwise segments and set memory segment selector to ZERO
+                column.resize(
+                    hasher.trace_len() + bitwise.trace_len() + memory.trace_len(),
+                    Felt::ZERO,
+                );
+                // set selector value for the final segment to ONE
+                column.resize(trace_len, Felt::ONE);
+                // add hasher segment to the hasher fragment to be filled from the hasher trace
+                let rest_of_column = hasher_fragment.push_column_slice(column, hasher.trace_len());
+                // add bitwise segment to the bitwise fragment to be filled from the bitwise trace
+                bitwise_fragment.push_column_slice(rest_of_column, bitwise.trace_len());
+            }
+            15 | 16 => {
+                // initialize hasher & memory segments and pad bitwise & final segments with ZERO
+                column.resize(trace_len, Felt::ZERO);
+                // add hasher segment to the hasher fragment to be filled from the hasher trace
+                let rest_of_column = hasher_fragment.push_column_slice(column, hasher.trace_len());
+                // split the column again to skip the bitwise segment, which has already been padded
+                let (_, rest_of_column) = rest_of_column.split_at_mut(bitwise.trace_len());
+                // add memory segment to the memory fragment to be filled from the memory trace
+                memory_fragment.push_column_slice(rest_of_column, memory.trace_len());
+            }
+            17 => {
+                // initialize hasher segment and pad bitwise, memory, and final segments with ZERO
+                column.resize(trace_len, Felt::ZERO);
+                // add hasher segment to the hasher fragment to be filled from the hasher trace
+                hasher_fragment.push_column_slice(column, hasher.trace_len());
+            }
+            _ => {
+                // initialize hasher, bitwise, memory segments and pad the final segment with ZERO
+                column.resize(trace_len, Felt::ZERO);
+                // add hasher segment to the hasher fragment to be filled from the hasher trace
+                let rest_of_column = hasher_fragment.push_column_slice(column, hasher.trace_len());
+                // add bitwise segment to the bitwise fragment to be filled from the bitwise trace
+                let rest_of_column =
+                    bitwise_fragment.push_column_slice(rest_of_column, bitwise.trace_len());
+                // add memory segment to the memory fragment to be filled from the memory trace
+                memory_fragment.push_column_slice(rest_of_column, memory.trace_len());
+            }
+        }
+    }
+
+    // fill the fragments with the execution trace from each coprocessor
+    hasher.fill_trace(&mut hasher_fragment);
+    bitwise.fill_trace(&mut bitwise_fragment);
+    memory.fill_trace(&mut memory_fragment);
```

Comment:
Ideally, we should fill these in parallel. This can be done in a future PR, but we should probably leave a `TODO` here or create an issue for it.

Reply:
TODO added!